File: internal/util/parse_size.go
package util

import (
	"fmt"
	"math"
	"regexp"
	"strconv"
	"strings"
)

var sizePattern = regexp.MustCompile(`^(\d+(?:\.\d+)?)\s*([BKMG])$`)

var sizeUnits = map[string]float64{
	"B": 1,
	"K": 1024,
	"M": 1024 * 1024,
	"G": 1024 * 1024 * 1024,
}

func ParseSize(value string) (int, error) {
	trimmed := strings.TrimSpace(value)
	match := sizePattern.FindStringSubmatch(strings.ToUpper(trimmed))
	if match == nil {
		return 0, fmt.Errorf("invalid size format: %q", value)
	}

	number, err := strconv.ParseFloat(match[1], 64)
	if err != nil {
		return 0, fmt.Errorf("invalid size value: %q", value)
	}

	multiplier, ok := sizeUnits[match[2]]
	if !ok {
		return 0, fmt.Errorf("unknown size unit: %q", match[2])
	}

	return int(math.Floor(number * multiplier)), nil
}

File: src/utils/parse-size.ts
const UNITS: Record<string, number> = {
  B: 1,
  K: 1024,
  M: 1024 * 1024,
  G: 1024 * 1024 * 1024,
};

export function parseSize(sizeStr: string): number {
  const match = /^(\d+(?:\.\d+)?)\s*([BKMG])$/i.exec(sizeStr.trim());
  if (!match) {
    throw new Error(
      `Invalid size format: "${sizeStr}". Expected format: <number><unit> (e.g., 1B, 2K, 2M)`,
    );
  }
  const value = Number(match[1]);
  const unit = match[2]!.toUpperCase();
  const multiplier = UNITS[unit];
  if (multiplier === undefined) {
    throw new Error(`Unknown size unit: "${unit}"`);
  }
  return Math.floor(value * multiplier);
}
